File: algorithms/dynamic-programming/coin-change/swift/CoinChange.swift
func coinChange(_ coins: [Int], _ amount: Int) -> Int {
    if amount == 0 { return 0 }

    var dp = Array(repeating: Int.max, count: amount + 1)
    dp[0] = 0

    for i in 1...amount {
        for coin in coins {
            if coin <= i && dp[i - coin] != Int.max {
                dp[i] = min(dp[i], dp[i - coin] + 1)
            }
        }
    }

    return dp[amount] == Int.max ? -1 : dp[amount]
}

print(coinChange([1, 5, 10, 25], 30)) // 2

File: algorithms/dynamic-programming/coin-change/rust/coin_change.rs
use std::cmp;

pub fn coin_change(coins: &[i32], amount: usize) -> i32 {
    if amount == 0 {
        return 0;
    }

    let mut dp = vec![i32::MAX; amount + 1];
    dp[0] = 0;

    for i in 1..=amount {
        for &coin in coins {
            let c = coin as usize;
            if c <= i && dp[i - c] != i32::MAX {
                dp[i] = cmp::min(dp[i], dp[i - c] + 1);
            }
        }
    }

    if dp[amount] == i32::MAX { -1 } else { dp[amount] }
}

fn main() {
    let coins = vec![1, 5, 10, 25];
    println!("{}", coin_change(&coins, 30)); // 2
}
